include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -1137,49 +1136,114 @@ struct __early_exit_find_or
+        auto __result_acc = __result_storage.__get_result_acc(__cgh);
+
+#if _ONEDPL_COMPILE_KERNEL && _ONEDPL_KERNEL_BUNDLE_PRESENT
+        __cgh.use_kernel_bundle(__kernel.get_kernel_bundle());
 #endif
-    auto __max_cu = oneapi::dpl::__internal::__max_compute_units(__exec);
+        __cgh.parallel_for<KernelName>(
+#if _ONEDPL_COMPILE_KERNEL && !_ONEDPL_KERNEL_BUNDLE_PRESENT
+            __kernel,
+#endif
+            sycl::nd_range</*dim=*/1>(sycl::range</*dim=*/1>(__n_groups * __wgroup_size),
+                                      sycl::range</*dim=*/1>(__wgroup_size)),
+            [=](sycl::nd_item</*dim=*/1> __item_id) {
+                auto __local_idx = __item_id.get_local_id(0);
+
+                // 1. Set initial value to local found state
+                _AtomicType __found_local = __init_value;
+
+                // 2. Find any element that satisfies pred
+                //  - after this call __found_local may still have initial value:
+                //    1) if no element satisfies pred;
+                //    2) early exit from sub-group occurred: in this case the state of __found_local will updated in the next group operation (3)
+                __pred(__item_id, __rng_n, __iters_per_work_item, __n_groups * __wgroup_size, __found_local,
+                       __brick_tag, __rngs...);
+
+                // 3. Reduce over group: find __dpl_sycl::__minimum (for the __parallel_find_forward_tag),
+                // find __dpl_sycl::__maximum (for the __parallel_find_backward_tag)
+                // or update state with __dpl_sycl::__any_of_group (for the __parallel_or_tag)
+                // inside all our group items
+                if constexpr (__or_tag_check)
+                    __found_local = __dpl_sycl::__any_of_group(__item_id.get_group(), __found_local);
+                else
+                    __found_local = __dpl_sycl::__reduce_over_group(__item_id.get_group(), __found_local,
+                                                                    typename _BrickTag::_LocalResultsReduceOp{});
+
+                // Set local found state value value to global state to have correct result
+                // - do not check __found_local != __init_value due __result_storage isn't initialized with __init_value on host side.
+                if (__local_idx == 0)
+                {
+                    _AtomicType& __found =
+                        *__result_and_scratch_storage_t::__get_usm_or_buffer_accessor_ptr(__result_acc);
+
+                    // Update global (for all groups) non-atomic state with the found index
+                    __found = __found_local;
+                }
+            });
+    });
 
-    auto __n_groups = oneapi::dpl::__internal::__dpl_ceiling_div(__rng_n, __wgroup_size);
-    __n_groups = ::std::min(__n_groups, decltype(__n_groups)(__max_cu));
+    // Wait and return result
+    return __future(__event_id, __result_storage).get();
```

Comment:
Why do we create a future here? Can't we just wait on the event and call `__get_value` on `__result_storage`?

Reply:
Who will call `sycl::queue::wait()` method in this case?

Reply:
The caller of `__get_value` needs to ensure that the result is ready. See the comment in https://github.com/oneapi-src/oneDPL/blob/825ba2f02fe8110e93afa1a49c7292d646851690/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h#L628-L629.

Reply:
How our __future::get() is implemented:
```C++
    auto
    get()
    {
        if constexpr (sizeof...(_Args) > 0)
        {
            auto& __val = std::get<0>(*this);
            return __wait_and_get_value(__val);
        }
        else
            wait();
    }
```
 
How __wait_and_get_value is implemented:
```C++
    template <typename _ExecutionPolicy, typename _T>
    constexpr auto
    __wait_and_get_value(__result_and_scratch_storage<_ExecutionPolicy, _T>& __storage)
    {
        if (__storage.is_USM())
            wait();
        return __storage.__get_value();
    }
```
 
So for me this mean that we really should call wait() only for USM memory.

@julianmi, am I correct?

Reply:
Yes, we only need explicit synchronization when using USM. I think we should remove the future since it is not used and change this line to `if (__result_storage.is_USM()) wait(); __result_storage.__get_value();`.

Reply:
I would suggest adding a `__result_and_scratch_storage.__wait_and_get_value` with this logic and use it here and in `__future.__wait_and_get_value` to not duplicate code.

Reply:
Should this new implementation wait under `ONEDPL_ALLOW_DEFERRED_WAITING` macro check like in the `__future` struct?

Reply:
> I would suggest adding a `__result_and_scratch_storage.__wait_and_get_value` with this logic and use it here and in `__future.__wait_and_get_value` to not duplicate code.

done.